c++/src/capnp/compiler/capnpc-c++.c++

```diff
@@ -1719,6 +1724,56 @@ private:
 
   // -----------------------------------------------------------------
 
+  enum class AsGenericRole { READER, BUILDER, CLIENT };
+
+  kj::StringTree makeAsGenericDef(AsGenericRole role, const TemplateContext& templateContext,
+                                  kj::StringPtr type, kj::String innerType = kj::String()) {
+    if (!templateContext.isGeneric()) {
+      return kj::StringTree();
+    }
+    kj::StringTree self, up;
+    if (templateContext.hasParams()) {
+      auto returnType = [&]() {
+        return kj::strTree(type, templateContext.args("2"), innerType);
+      };
+      kj::StringTree asGeneric = kj::strTree("as", innerType.size() ? type : "", "Generic()");
+      switch (role) {
+        case AsGenericRole::READER:
+          self = kj::strTree(
+              "  ", templateContext.decl(true, "2"),
+              "  typename ", returnType(), "::Reader ", kj::mv(asGeneric), " {\n"
+              "    return typename ", returnType(), "::Reader(_reader);\n"
+              "  }\n"
+              "\n");
+          break;
+        case AsGenericRole::BUILDER:
+          self = kj::strTree(
+              "  ", templateContext.decl(true, "2"),
+              "  typename ", returnType(), "::Builder ", kj::mv(asGeneric), " {\n"
+              "    return typename ", returnType(), "::Builder(_builder);\n"
+              "  }\n"
+              "\n");
+          break;
+        case AsGenericRole::CLIENT:
+          self = kj::strTree(
+              "  ", templateContext.decl(true, "2"),
+              "  typename ", returnType(), "::Client ", kj::mv(asGeneric), " {\n"
+              "    return castAs<", innerType.size() ? "typename " : "", returnType(), ">();\n"
+              "  }\n"
+              "\n");
+          break;
+      }
+    }
+    KJ_IF_MAYBE(p, templateContext.getParent()) {
+      up = makeAsGenericDef(role, *p, p->getName(), kj::strTree(
+          templateContext.hasParams() ? "::template " : "::", type,
+          templateContext.args()).flatten());
```

Comment:
Oh I see, this is the argument to makeAsGenericDef(). Perhaps the `innerType` parameter should be type `StringTree`?